### ISSUE TITLE
Maybe consider getting rid of some external dependencies

### ISSUE THREAD

**Issue by tulilirockz** [ublue-os/bluefin#2039]
I don't want to be that security-focused weirdo person that thinks everything is a huuuuge booggie man and is gonna destroy the project, but I feel like as this project is getting more and more traction, we should consider reacessing which external dependencies are being added at least for the base Bluefin images (non -dx).

I'd consider an "external dependency" something that does not come from either the [Fedora official repositories](https://packages.fedoraproject.org/) (rpm-fusion too), [Homebrew packages](https://formulae.brew.sh/), [Flathub](https://flathub.org/), or our own [packages](https://github.com/ublue-os/packages) repository (or anything else on our org, like akmods, and others). Considering that, we currently have these external dependencies on base:

- https://copr.fedorainfracloud.org/coprs/sentry/switcheroo-control_discrete
- https://copr.fedorainfracloud.org/coprs/che/nerd-fonts
- https://github.com/starship/starship
- https://raw.githubusercontent.com/rcaloras/bash-preexec
- Mediatek firmware from https://gitlab.com/kernel-firmware/linux-firmware
- [Charm Repo](https://github.com/ublue-os/bluefin/blob/main/system_files/shared/etc/yum.repos.d/charm.repo)
- [Tailscale Repo](https://github.com/ublue-os/bluefin/blob/main/system_files/shared/etc/yum.repos.d/tailscale.repo)

On HWE:
- https://copr.fedorainfracloud.org/coprs/lukenukem/asus-linux
- https://github.com/linux-surface/linux-surface patches

On DX we have these:
- https://copr.fedorainfracloud.org/coprs/ganto/lxc4
- https://copr.fedorainfracloud.org/coprs/ganto/umoci
- https://copr.fedorainfracloud.org/coprs/karmab/kcli
- https://copr.fedorainfracloud.org/coprs/atim/ubuntu-fonts
- https://copr.fedorainfracloud.org/coprs/hikariknight/looking-glass-kvmfr
- https://copr.fedorainfracloud.org/coprs/gmaglione/podman-bootc
- https://github.com/kubernetes-sigs/kind
- https://github.com/githubnext/monaspace

Other than these there are just things getting pulled from [hwe](https://github.com/ublue-os/hwe) and [akmods](https://github.com/ublue-os/akmods), nothing really to worry about. But I'd just like to ask you guys if there would be anything that would be best to be moved to [packages](https://github.com/ublue-os/packages) so that we could have more control over, or just straight up removed from the system

**Comment by tulilirockz** [comment]
I believe the HWE ones are gonna get removed at some point, arent they? - Otherwise honestly I think just stuff like the `ubuntu-fonts` copr is kind of redundant since we have them on-repo already

**Comment by tulilirockz** [comment]
I believe a improvement would be a way to know which coprs are being used for which packages on the `packages.json`. Currently I believe someone could for example make a custom package w/ the name of another one using an existing copr to inject other programs into the build for example. I know this is being paranoid but its still kinda concerning

**Comment by castrojo** [comment]
This has been on the todo for a while so might as well take care of it. The HWE ones go away, I'm fine with deps on the upstreams like k8s, tailscale, starship, etc. Bazitte uses Switcheroo and we work with sentry already, though it might make sense to move that one to ublue-os/packages since bazzite uses it. Hikari is on the ublue core team. The podman-bootc one is run by a redhat and fedora person, and likely has a path into Fedora anyway. 

The Incus stuff either gets better in-distro or we may choose to use timothee's sysexts for that. 

I feel like fonts should be a separate audit because that's a whole ball of stuff. I forgot what bash-preexec does, I think we use it with distrobox?

**Comment by m2Giles** [comment]
Bash pre exec is for atuin.

**Comment by klmcwhirter** [comment]
At the very least, I would like to have more 'toggle-' options in ujust until the strategy for external dependencies have been worked through.

I am currently coming to the conclusion that starship is a buggy and limiting mess. I am not sure what I am going to replace it with yet. But I would like to turn it off. For now I plan on just commenting out the line at the bottom of `/etc/bashrc`.

I would rather do `ujust toggle-starship` so I don't have to learn what other places in the system were modified to fully integrate it. I hope it is just the one place, but ...

That was just an example. In general, I would suggest something similar for the rest - within reason.

**Comment by castrojo** [comment]
https://docs.projectbluefin.io/FAQ#starship-is-not-for-me-how-do-i-disable-it

**Comment by klmcwhirter** [comment]
> https://docs.projectbluefin.io/FAQ#starship-is-not-for-me-how-do-i-disable-it

Yeah, I did that and it seemed to work.

My point is that until things can get to a point where all external dependencies are opt-in (as @tulilirockz points out in the original purpose of the request) - there are too many places to discover things like this. FAQ, other places in the docs or upstream docs and ujust.

ujust has these:
- toggle-nvk
- toggle-updates
- toggle-user-motd
- toggle-devmode
- toggle-tailscale

What I am suggesting is pick one way of communicating optional system modifications and stick with it.

So adding this seems like a good step forward:
- toggle-starship

That way if you need to modify how starship is integrated (perhaps by adding an opinionated default starship.toml file) then you could just modify the justfile and deal with it there. No need to document those *changing* details.

The users of bluefin will thank you. And you will get less questions about these things.

**Comment by castrojo** [comment]
You don't need permission, send a pull request and we can take a look!

**Comment by klmcwhirter** [comment]
I was just participating in the discussion. Seems this request is in an early design phase.

Once an approach has been settled on, sure I would be willing to help. Development, docs, testing, whatever.

----

I am loving bluefin! It is really causing me to challenge how I have been using my system. I am truly enjoying finding new techniques.

For example, I use VS Code (devcontainers) and distrobox pets. But as I posted in the UB forum, they just do not play nice with each other.

My current experiment has me building a set of OCI images based on `ghcr.io/ublue-os/fedora-toolbox:latest` that meets the needs of both.

Then I use those common images for both distrobox and devcontainers. There is some complexity that I am working through. But there are some interesting benefits. For example, topgrade seems to update my running decontainers as well as distrob-oxen (pet pun intended). That is a cool side effect.

It is early on, but the approach seems promising ...

**Comment by castrojo** [comment]
Ok let's do an update on where we stand on this.

**Comment by klmcwhirter** [comment]
Thanks for the work you folks are doing to clean the image definition and build process. Us users (er, abusers) will all benefit.
> Thanks to those of you who have tolerated my questions as I have been coming up to speed.

I don't have anything specific to contribute, but let me share my perspective as a dx user.

I am a developer. I work with a lot of different tool chains but I really like bluefin - because NVIDIA. I am happy to defer to those who are tracking those kernel changes closely to keep my system running smooth. I prefer to spend my time writing apps and not troubleshooting hardware. I think I have spent enough decades of my life doing that.

And so for that one main reason I like that bluefin is opinionated - so I don't have to have an opinion.

But there are quite a few other opinions that exist in bluefin that I would like to have an easy way to express my own where I have one. And to be able to discover them simply without resorting to asking questions.

The use of `ujust` is one way; managing apps via flatpak and brew are another. As long as the approach being used does not involve too much documentation it should be fine. Some docs are ok. But if they need to be maintained then that adds friction to the maintenance team moving quickly.

Again thanks for what you are doing. I am still learning. But, wow, what an environment to do development in. Switching to a whole new version of a complex toolchain can be as simple as just using a different distrobox.

I recently moved to python 3.14.0a6 by compiling from source. I built a new distrobox to build it and test some of the surrounding tools I needed. When I saw issues, I just removed the image and tried again later when the issues were addressed. Once they were close enough I committed to the version by installing to my host $HOME/.local dir structure. 

Simply awesome.

**Comment by castrojo** [comment]
>  As long as the approach being used does not involve too much documentation it should be fine. Some docs are ok.

Yeah this is on purpose, if we have to document a thing too much it means we're special, so we purposely try to focus on the upstream tooling, thanks for your feedback!

**Comment by klmcwhirter** [comment]
@castrojo Well, I happen to think Universal Blue is something special. ;)

You guys are pushing the notion of Solution Integrator to a new place. And providing some real value adds along the way.

There is a lot of promise in what you are doing.

And in addition, you have forced this old dog to learn some new tricks.

This is what innovation looks like folks!
